File: test-suite-python/src/test/python/micronaut/docs/hibernate/graalvm/OrderRepository.py
from typing import Annotated

from jakarta.inject import Singleton
from jakarta.persistence import EntityManager, PersistenceContext
from micronaut.transaction.annotation import Transactional

from micronaut.docs.hibernate.graalvm.Order import Order
from micronaut.docs.hibernate.graalvm.OrderId import OrderId


@Singleton
class OrderRepository:
    entity_manager: Annotated[EntityManager, PersistenceContext]

    @Transactional
    def save(self, id: OrderId, customer: str) -> Order:
        order = Order()
        order.id = id
        order.customer = customer
        self.entity_manager.persist(order)
        return order

    @Transactional(readOnly=True)
    def find_by_id(self, id: OrderId) -> Order | None:
        return self.entity_manager.find(Order, id)

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/hibernate/graalvm/OrderRepository.kt
package io.micronaut.docs.hibernate.graalvm

import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Singleton
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext

@Singleton
open class OrderRepository {

    @PersistenceContext
    lateinit var entityManager: EntityManager

    @Transactional
    open fun save(id: OrderId, customer: String): Order {
        val order = Order()
        order.id = id
        order.customer = customer
        entityManager.persist(order)
        return order
    }

    @Transactional(readOnly = true)
    open fun findById(id: OrderId): Order? = entityManager.find(Order::class.java, id)
}
